### DISCUSSION TITLE
Custom equality policy / hashing strategy for keys

### DISCUSSION THREAD

**Discussion by batya239** [ben-manes/caffeine#856] [Ideas]
Hi, Ben!

Currently caffeine relies only for canonical equality/hashing policy via proper defined #equals() and #hashCode() methods. Do you have any plans to support something like 

```
public interface HashingStrategy<K> {
  int hashCode(K key);

  boolean equals(K k1, K k2);
}
```

? 

Or would you mind for pull request to support it?

**Comment by ben-manes** [comment]
There is no planned support. It would fall under two strategies, which at this time don't make sense for us.

### Wrap the key

The key can be wrapped with a custom equivalence strategy, like with any other hashmap type. As this can easily be done in user code, it doesn't carry its weight to make as a feature in this library. See Guava's [Equivalence](https://guava.dev/releases/31.1-jre/api/docs/com/google/common/base/Equivalence.html) and the `wrap` method.

### Custom hash table

The wrapper can be avoided by having the hash table call the equivalence strategy's methods with the key. However, this requires forking `ConcurrentHashMap` to rewrite its usages, which is not desirable. It would require that we keep in sync with a complex, evolving class which uses low-level feature only available to the JVM authors. However, if the Java authors decided to support this then we would expose the plumbing (you could ask on [core-libs-dev](https://mail.openjdk.org/pipermail/core-libs-dev/)).
